core/trino-main/src/main/java/io/trino/json/ir/SqlJsonLiteralConverter.java

```diff
@@ -0,0 +1,182 @@
+                    return Optional.of(new TypedValue(INTEGER, jsonNode.longValue()));
+                }
+                if (jsonNode.canConvertToLong()) {
+                    return Optional.of(new TypedValue(BIGINT, jsonNode.longValue()));
+                }
+                throw conversionError(jsonNode, "value too big");
+            }
+            if (jsonNode instanceof DecimalNode) {
+                BigDecimal jsonDecimal = jsonNode.decimalValue();
+                int precision = jsonDecimal.precision();
+                if (precision > MAX_PRECISION) {
+                    throw conversionError(jsonNode, "precision too big");
+                }
+                int scale = jsonDecimal.scale();
+                DecimalType decimalType = createDecimalType(precision, scale);
+                Object value = decimalType.isShort() ? encodeShortScaledValue(jsonDecimal, scale) : encodeScaledValue(jsonDecimal, scale);
+                return Optional.of(TypedValue.fromValueAsObject(decimalType, value));
+            }
+            if (jsonNode instanceof DoubleNode) {
+                return Optional.of(new TypedValue(DOUBLE, jsonNode.doubleValue()));
+            }
+            if (jsonNode instanceof FloatNode) {
+                return Optional.of(new TypedValue(REAL, floatToRawIntBits(jsonNode.floatValue())));
+            }
+            if (jsonNode instanceof IntNode) {
+                return Optional.of(new TypedValue(INTEGER, jsonNode.longValue()));
+            }
+            if (jsonNode instanceof LongNode) {
+                return Optional.of(new TypedValue(BIGINT, jsonNode.longValue()));
+            }
+            if (jsonNode instanceof ShortNode) {
+                return Optional.of(new TypedValue(SMALLINT, jsonNode.longValue()));
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    public static Optional<JsonNode> getJsonNode(TypedValue typedValue)
+    {
+        Type type = typedValue.getType();
+        if (type.equals(BOOLEAN)) {
+            return Optional.of(BooleanNode.valueOf(typedValue.getBooleanValue()));
+        }
+        if (type instanceof CharType) {
+            return Optional.of(TextNode.valueOf(padSpaces((Slice) typedValue.getObjectValue(), (CharType) typedValue.getType()).toStringUtf8()));
+        }
+        if (type instanceof VarcharType) {
+            return Optional.of(TextNode.valueOf(((Slice) typedValue.getObjectValue()).toStringUtf8()));
+        }
+        if (type.equals(BIGINT)) {
+            return Optional.of(LongNode.valueOf(typedValue.getLongValue()));
+        }
+        if (type.equals(INTEGER)) {
+            return Optional.of(IntNode.valueOf((int) typedValue.getLongValue()));
+        }
+        if (type.equals(SMALLINT)) {
+            return Optional.of(ShortNode.valueOf((short) typedValue.getLongValue()));
+        }
+        if (type.equals(TINYINT)) {
+            return Optional.of(ShortNode.valueOf((short) typedValue.getLongValue()));
```

Comment:
would it make sense to use `toIntExact` and similar here, for extra safety that we are not loosing precision.